crates/sui-indexer/src/type_tag_parser.rs

```diff
@@ -0,0 +1,470 @@
+        Token::U128Type => "u128".to_string(),
+        Token::U256Type => "u256".to_string(),
+        Token::BoolType => "bool".to_string(),
+        Token::AddressType => "address".to_string(),
+        Token::VectorType => "vector".to_string(),
+        Token::True => "true".to_string(),
+        Token::False => "false".to_string(),
+        Token::SignerType => "signer".to_string(),
+        Token::Name(name) => name,
+        _ => bail!("Unexpected token {:?} expected a name", t),
+    })
+}
+
+fn name_token(s: String) -> Token {
+    match s.as_str() {
+        "u8" => Token::U8Type,
+        "u16" => Token::U16Type,
+        "u32" => Token::U32Type,
+        "u64" => Token::U64Type,
+        "u128" => Token::U128Type,
+        "u256" => Token::U256Type,
+        "bool" => Token::BoolType,
+        "address" => Token::AddressType,
+        "vector" => Token::VectorType,
+        "true" => Token::True,
+        "false" => Token::False,
+        "signer" => Token::SignerType,
+        _ => Token::Name(s),
+    }
+}
+
+fn next_number(initial: char, mut it: impl Iterator<Item = char>) -> Result<(Token, usize)> {
+    let mut num = String::new();
+    num.push(initial);
+    loop {
+        match it.next() {
+            Some(c) if c.is_ascii_digit() || c == '_' => num.push(c),
+            Some(c) if c.is_alphanumeric() => {
+                let mut suffix = String::new();
+                suffix.push(c);
+                loop {
+                    match it.next() {
+                        Some(c) if c.is_ascii_alphanumeric() => suffix.push(c),
+                        _ => {
+                            let len = num.len() + suffix.len();
+                            let tok = match suffix.as_str() {
+                                "u8" => Token::U8(num),
+                                "u16" => Token::U16(num),
+                                "u32" => Token::U32(num),
+                                "u64" => Token::U64(num),
+                                "u128" => Token::U128(num),
+                                "u256" => Token::U256(num),
+                                _ => bail!("invalid suffix"),
+                            };
+                            return Ok((tok, len));
+                        }
+                    }
+                }
+            }
+            _ => {
+                let len = num.len();
+                return Ok((Token::U64(num), len));
+            }
+        }
+    }
+}
+
+#[allow(clippy::many_single_char_names)]
+fn next_token(s: &str) -> Result<Option<(Token, usize)>> {
+    let mut it = s.chars().peekable();
+    match it.next() {
+        None => Ok(None),
+        Some(c) => Ok(Some(match c {
+            '<' => (Token::Lt, 1),
+            '>' => (Token::Gt, 1),
+            ',' => (Token::Comma, 1),
+            ':' => match it.next() {
+                Some(':') => (Token::ColonColon, 2),
+                _ => bail!("unrecognized token"),
```

Comment:
should we be a bit more specific for the errors?
"missing :" or "':' expected"
same for other bail